### DISCUSSION TITLE
NoiseModelling now use Tinfour

### DISCUSSION THREAD

**Discussion by pierromond** [gwlucastrig/Tinfour#72] [Show and tell]
Thank you very much ! @gwlucastrig  @micycle1 . We now use Tinfour in NoiseModelling and this has greatly improved the performance of the triangulation phase of our tool : https://github.com/Ifsttar/NoiseModelling/pull/331

Cheers !

**Comment by nicolas-f** [comment]
Hello,
We have written a class layer in order to use JTS for input/output geometries. We use Quadtree in order to merge close vertices

https://github.com/Ifsttar/NoiseModelling/blob/v3.4/noisemodelling-pathfinder/src/main/java/org/noise_planet/noisemodelling/pathfinder/LayerTinfour.java

**Reply by gwlucastrig** [reply]
I looked at your code.  It appears to use Tinfour well. Are there any features that I could add that would be helpful in the future?

How many vertices does your application use when you run a noise model?

Also, there is a new iterator for SimpleTriangles in the Incremental TIN classes that you may find helpful for future work.

    for(SimpleTriangle triangle: tin.triangles() ){
     }

Because Tinfour is intended to support very large data sets, memory conservation is an important part of its design. Thus, SimpleTriangles are not maintained as objects inside the IncrementalTin instances, but are produced dynamically as needed.

**Reply by nicolas-f** [reply]
The typical area of computation is 1 square km . With a typical digital elevation model of 25m and currently with buildings and roads also. So not much more than 1 million vertices. We are currently more involved into robustness than memory usage.

A missing feature (or not easy to deduce) was to create an array of SimpleTriangle and finding the immediate neighbour SimpleTriangles.

**Reply by micycle1** [reply]
> A missing feature (or not easy to deduce) was to create an array of SimpleTriangle and finding the immediate neighbour SimpleTriangles.

My solution for this is to first build a HashMap of Edge -> Triangle (since each triangle has 3 edges, insert a mapping for each edge) during `TriangleCollector.visitSimpleTriangles()`. Then the neighbours for a given triangle are:

```
final SimpleTriangle n1 = map.get(t.getEdgeA().getDual());
final SimpleTriangle n2 = map.get(t.getEdgeB().getDual());
final SimpleTriangle n3 = map.get(t.getEdgeC().getDual());
```

**Reply by gwlucastrig** [reply]
Would it be helpful if SimpleTriangle included a getNeighbor() method of some sort?  Maybe something like:

    for(SimpleTriangle trig: tin.triangles()){
           SimpleTriangle a = trig.getNeighborA();    // neighbor triangle sharing edge a.
           SimpleTriangle b = trig.getNeighborB();
           SimpleTriangle c = trig.getNeighborC();
    }

**Reply by nicolas-f** [reply]
Yes I think, if the HashCode of the SimpleTriangle is the same for the same set of edges

**Comment by gwlucastrig** [comment]
You are quite welcome. I am pleased to see Tinfour being used by such an interesting project. 

Also, I saw that you awarded Tinfour a star on Github a couple of weeks ago. Thank you very much for that.

**Comment by gwlucastrig** [comment]
It sounds to me like there's a couple of things that SimpleTriangle could probably use.   Perhaps a getIndex() method that gets a unique index for each triangle, an equals() method, and a hash() method.   And, of course, a get neighbor method.  I will look into these.

My timing here isn't great because a Tinfour release was just posted a couple of weeks ago and it's going to be a few months before I have another one ready.  But the code will, of course, be posted at the Github repository in the mean time.
